### PR TITLE
Taskcluster publication

### PR DESCRIPTION
**Pull request by La0**
Refs #333

This PR add code to run `parse_tc_logs` using live logs from teacher trainer task.

The call to the python script is configured through a Taskcluster transform `use-secret` which brings
- Taskcluster proxy
- Taskcluster secret scope
- Taskcluster secret environment variable used by `parse_tc_logs`

When the secret is available and the `parse_tc_logs` has been installed through `pip install ./training`, the command will run against the logs to track & publish metrics.

If the command crashes (invalid or missing credentials for example), it will cleanly exit with exit code 0 to prevent crashing the training too.

Taskcluster header parsing has been disabled when using `--from-stream` as these headers are NOT present when running from inside the task.

### CODE REVIEW

**Comment by La0** [comment]
Got a [training task](https://firefox-ci-tc.services.mozilla.com/tasks/OZc_WFtKSc2MaNxq9OUOFg/definition) definition with 
- taskcluster proxy enabled
- tracking code install through pip
- taskcluster secret env variable set

File: pipeline/train/train.sh
Comment (bhearsum) [inline pipeline/train/train.sh:142]: question: Do we actually want to combine stderr and stdout? I imagine the parser handles it fine, but I'm not sure if it has downsides for the `tee` case.

File: taskcluster/translations_taskgraph/parameters.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/parameters.py:46]: suggestion: We probably don't want to take this change (I think you needed it for testing though).

(I'll defer to @eu9ene on whether or not this is blocking.)

File: taskcluster/translations_taskgraph/transforms/worker_env.py
Comment (bhearsum) [inline taskcluster/translations_taskgraph/transforms/worker_env.py:83]: suggestion: please put this in the `kind` instead. We've been trying to keep as much as possible in there to improve readability. It does increase repetition, but it makes it easier to map tasks back to kinds and make tweaks. (If you feel this must go in transform, please use a new one -- this one is _specifically_ for dealing with worker-type specific configuration, which this is not.)

**Review by bhearsum — CHANGES_REQUESTED** [review]
I only reviewed the taskcluster parts. Let me know if you have any questions!

File: taskcluster/translations_taskgraph/parameters.py
Comment (eu9ene) [inline taskcluster/translations_taskgraph/parameters.py:48]: This should be reverted, validation runs correctly once in a regular CI run

File: tracking/translations_parser/parser.py
Comment (eu9ene) [inline tracking/translations_parser/parser.py:196]: Should this be removed? I think the idea was to just propagate the existing logging to stdout as.

**Review by eu9ene — CHANGES_REQUESTED** [review]
I compared the [new train-teacher log](https://firefox-ci-tc.services.mozilla.com/tasks/OCGS_xxgQWSYqbJ09aFw2Q/runs/0/logs/public/logs/live.log) form CI with the[ previous one](https://firefox-ci-tc.services.mozilla.com/tasks/VGzXs7RnQWeaqfp5hhCKeg/runs/0/logs/public/logs/live.log). The wrapper should not modify the original log. It should stream stdout and stderr of the wrapped program as is and optionally add some logging records with the same convention. For example: `[tracking] [INFO] Reading logs stream.`

Also, it complains about missing validation here but in the original log we can see validation steps happing correctly:
```
[task 2024-04-15T15:48:33.948Z] [WARNING] Some validation data is incomplete with the following epoch/up couples:
[task 2024-04-15T15:48:33.948Z] [WARNING] * Ep. 1, Up. 49
[task 2024-04-15T15:48:33.948Z] [WARNING] * Ep. 1, Up. 50
[task 2024-04-15T15:48:33.949Z] [WARNING] No ValidationEpoch entry, skipping.
```

I did not find a test publication in https://wandb.ai/moz-translations/ru-en.

File: taskcluster/kinds/train-teacher/kind.yml
Comment (eu9ene) [inline taskcluster/kinds/train-teacher/kind.yml:107]: Let's also add it to all other steps that use `train-taskcluster.sh` (train-*, finetune-*)

**Comment by bhearsum** [comment]
> I compared the [new train-teacher log](https://firefox-ci-tc.services.mozilla.com/tasks/OCGS_xxgQWSYqbJ09aFw2Q/runs/0/logs/public/logs/live.log) form CI with the[ previous one](https://firefox-ci-tc.services.mozilla.com/tasks/VGzXs7RnQWeaqfp5hhCKeg/runs/0/logs/public/logs/live.log). The wrapper should not modify the original log. It should stream stdout and stderr of the wrapped program as is and optionally add some logging records with the same convention. For example: [tracking] [INFO] Reading logs stream.

I wonder if this is related to my comment about stdout & stderr getting combined? I imagine that separating those (and handling both streams) might result in better behaviour. Even so - I'm not certain whether or not a downstream consumer can guarantee that they will process messages from two separate pipes in the same order they would've been printed to the console. (Maybe it is possible...I honestly don't know.)

**Comment by eu9ene** [comment]
> > I compared the [new train-teacher log](https://firefox-ci-tc.services.mozilla.com/tasks/OCGS_xxgQWSYqbJ09aFw2Q/runs/0/logs/public/logs/live.log) form CI with the[ previous one](https://firefox-ci-tc.services.mozilla.com/tasks/VGzXs7RnQWeaqfp5hhCKeg/runs/0/logs/public/logs/live.log). The wrapper should not modify the original log. It should stream stdout and stderr of the wrapped program as is and optionally add some logging records with the same convention. For example: [tracking] [INFO] Reading logs stream.
> 
> I wonder if this is related to my comment about stdout & stderr getting combined? I imagine that separating those (and handling both streams) might result in better behaviour. Even so - I'm not certain whether or not a downstream consumer can guarantee that they will process messages from two separate pipes in the same order they would've been printed to the console. (Maybe it is possible...I honestly don't know.)

We discussed live that it's ok to merge those two streams because they are intentionally used by OpusTrainer to output stuff. I guess it was used to separate Marian log and OpusTrainer's own log. Unless there's a way to easily redirect both separately using bash piping.

**Comment by eu9ene** [comment]
Some other things we discussed should be implemented before merging:
- we should investigate warnings with validation publishing. Indeed I don't see validation graphs on https://wandb.ai/moz-translations/moz-translations
- we should add a parameter to the training config to disable publishing on CI and when we debug something and don't want to pollute the dashboards
- as in the previous comment, preserving log format and adding [tracking] prefix for the parser
- using the proper project/experiment name for publishing right away as we'll start using this when it's merged (should be `ru-en/ci_<task_cluster_id>` for the test CI run when enabled
- making sure all the training and fine-tuning steps are tracked

**Comment by bhearsum** [comment]
> Some other things we discussed should be implemented before merging:
>     * we should add a parameter to the training config to disable publishing on CI and when we debug something and don't want to pollute the dashboards

Apologies that I missed the meeting. One possible alternative here, if you want, might be to have a separate publishing location (some sort of sandbox) that we use for these cases.

**Comment by eu9ene** [comment]
> > Some other things we discussed should be implemented before merging:
> > * we should add a parameter to the training config to disable publishing on CI and when we debug something and don't want to pollute the dashboards
> 
> Apologies that I missed the meeting. One possible alternative here, if you want, might be to have a separate publishing location (some sort of sandbox) that we use for these cases.

It's a good idea. We could create a staging namespace for testing, I just don't know if we want to always publish there. The namespace can be called something like `moz-translations-staging` and then we can specify the namespace in the config. An empty value can mean that we don't want to publish anywhere. @La0 we can use an environment variable for this. It will be easier than propagating the argument through many layers. This namespace will be our sandbox for testing before rolling out to production. If everyone is ok with this approach I'll ask the MLOps team to create it for us. cc @gregtatum

**Comment by eu9ene** [comment]
> > > Some other things we discussed should be implemented before merging:
> > > 
> > > * we should add a parameter to the training config to disable publishing on CI and when we debug something and don't want to pollute the dashboards
> > 
> > 
> > Apologies that I missed the meeting. One possible alternative here, if you want, might be to have a separate publishing location (some sort of sandbox) that we use for these cases.
> 
> It's a good idea. We could create a staging namespace for testing, I just don't know if we want to always publish there. The namespace can be called something like `moz-translations-staging` and then we can specify the namespace in the config. An empty value can mean that we don't want to publish anywhere. This namespace will be our sandbox for testing before rolling out to production. If everyone is ok with this approach I'll ask the MLOps team to create it for us. cc @gregtatum
> 
> I think in the short term we can implement it and put an empty string into the CI config. Then when the new namespace is created we'll be able to retest the full reuploading of the old experiments there instead of Teklia project.

UPD: I looked at the code and I see that what W&B namespace we write to is controlled by the secret token, so to implement this we would need to deal with secrets. Let's leave it out of scope for this PR and just make sure we can disable publishing from the experiment config.

File: taskcluster/translations_taskgraph/transforms/worker_env.py
Comment (La0) [reply taskcluster/translations_taskgraph/transforms/worker_env.py]: I moved these settings to kind in 32f5c4e

File: taskcluster/translations_taskgraph/parameters.py
Comment (La0) [reply taskcluster/translations_taskgraph/parameters.py]: I only set this to force-trigger a training task.

I explicitly added a `[TRASHME]` in the commit message to avoid merging the PR with this commit

File: tracking/translations_parser/parser.py
Comment (La0) [reply tracking/translations_parser/parser.py]: I replaced this by outputting the raw line unmodified, and prefixing log lines from parser with `[tracking]`

**Comment by eu9ene** [comment]
> @eu9ene Is this expected behaviour ? Can we set `perplexity` as optional ?

@La0 All the metrics that are used for validation are set in the Marian config (--valid-metrics), so they are all kind of optional. We usually don't use perplexity. We set them here https://github.com/mozilla/firefox-translations-training/blob/145a84ace322184b5c5d6e9aca0ececb037c5d08/pipeline/train/train.sh#L40

**Comment by La0** [comment]
Last changes:
- I made `perplexity` optional
- A training parameter `wandb-publication` has been added, disabled by default for CI tasks. You can control it when starting a new training

File: pipeline/train/train.sh
Comment (eu9ene) [inline pipeline/train/train.sh:112]: @La0 naming should be consistent with the offline uploading. For group: `<experiment-name-from-training-config>_<taskcluster-group-id>`, for task: I think we parse the task label to get one of the following: finetune-student, student, teacher-1, teacher-2, backwards. Some examples:

Screenshot 2024-04-22 at 2 26 36 PM
Screenshot 2024-04-22 at 2 22 10 PM

Comment (eu9ene) [inline pipeline/train/train.sh:112]: It might be better to move this code to a transform where we can properly retrieve the data like Taskcluster task group id and the experiment ID for the CLI arguments. Ideally, this task should not depend on Taskcluster as we usually separate task implementation and the orchestrator-specific code. This job can potentially be launched locally or using a different orchestrator( snakemake etc.)

Comment (La0) [reply pipeline/train/train.sh:112]: @bhearsum was against transform usage (see comment [here](https://github.com/mozilla/firefox-translations-training/pull/501#discussion_r1565928735)).

I'll probably move the naming arguments into the python code, as it will be easier to condition these parameters for taskcluster usage

Comment (La0) [reply pipeline/train/train.sh]: The naming is now fully handled by the tracking Python code, detecting if we are running in Taskcluster, in CI or normal runs, and producing the names using the group publication algorithm.

I setup default values for CI as the configuration is not exposed through task group in that case

**Review by bhearsum — APPROVED** [review]
Looks great from a taskgraph standpoint!

File: pipeline/train/train.sh
Comment (eu9ene) [reply pipeline/train/train.sh]: moving to the publisher is a good idea!

Comment (eu9ene) [reply pipeline/train/train.sh]: - [x] The group name looks good
- [x] The graphs look good
- [ ] The run names are still not consistent with the offline publisher

It's important to make naming consistent to compare the same model across older and newer training. See the names in en-ca: https://wandb.ai/moz-translations/en-ca

Comment (eu9ene) [reply pipeline/train/train.sh]: When this last issue is fixed, I'll run a train action to test how it displays graphs in real time and that it takes the values properly from the config

Comment (La0) [reply pipeline/train/train.sh]: Done in [37d227f](https://github.com/mozilla/firefox-translations-training/pull/501/commits/37d227f52bc7339e51d8fe6430357f1e2b47ca9a)

**Comment by eu9ene** [comment]
I launched a test training run from this branch https://firefox-ci-tc.services.mozilla.com/tasks/SngAEYS-RyWnY4GgKJTzSQ

**Comment by La0** [comment]
> I launched a test training run from this branch https://firefox-ci-tc.services.mozilla.com/tasks/SngAEYS-RyWnY4GgKJTzSQ

I do not see any training task in the group, maybe the cache did not allow your tasks to run ?

**Comment by eu9ene** [comment]
> > I launched a test training run from this branch https://firefox-ci-tc.services.mozilla.com/tasks/SngAEYS-RyWnY4GgKJTzSQ
> 
> I do not see any training task in the group, maybe the cache did not allow your tasks to run ?

It's running https://firefox-ci-tc.services.mozilla.com/tasks/PDAfVzlWQMam4uqhndjafg. You're probably looking at the actions group which is different from the tasks group. Also, I ran it from a push task.

It looks working! https://wandb.ai/moz-translations/en-ru/groups/test_tracking_SngAEYS-RyWnY4GgKJTzSQ

**Comment by La0** [comment]
I rebased on latest `main` commits, and removed the extra commit that tweaked training parameters.

File: tracking/translations_parser/cli/taskcluster.py
Comment (eu9ene) [inline tracking/translations_parser/cli/taskcluster.py:191]: I see that the tag "cli" is visible in the workgroup. I see we also use the tag "Taskcluster" in the code. Should we utilize tags to distinguish between different modes of publishing? We have 3 now: GCP, taskcluster-offline and taskcluster online.

**Review by eu9ene — APPROVED** [review]
I have a couple of questions/suggestions, otherwise it's good to go

File: tracking/translations_parser/cli/taskcluster.py
Comment (La0) [reply tracking/translations_parser/cli/taskcluster.py]: I exposed the tags through a new CLI argument `--tags`:
- on CI uploads, it will use `taskcluster`
- on offline group upload using the dedicated entry point, it will use `taskcluster-offline`
- when uploading a local folder, the operator should set `GCP`

File: tracking/translations_parser/cli/taskcluster.py
Comment (La0) [reply tracking/translations_parser/cli/taskcluster.py:135]: What should be the expected behaviour when running on CI without an explicit configuration ? We still need to produce weight & biases project, group & run names.

Comment (eu9ene) [reply tracking/translations_parser/cli/taskcluster.py:135]: as discussed we can keep it for testing for now

**Comment by eu9ene** [comment]
Ok, it looks good. If we find more issues or have more ideas on what to improve let's handle them in the follow-up tasks. We'll need to reupload the experiments anyway when it's feature complete. We need to make sure the publishing from CI is disabled though not to pollute the ru-en project.

**Pull request merged**
